swarm_copy/app/routers/tools.py

```diff
@@ -0,0 +1,112 @@
+"""Conversation related CRUD operations."""
+
+import json
+import logging
+from typing import Annotated, Any
+
+from fastapi import APIRouter, Depends, HTTPException
+from sqlalchemy import select
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from swarm_copy.app.database.db_utils import get_thread
+from swarm_copy.app.database.schemas import ToolCallSchema
+from swarm_copy.app.database.sql_schemas import Messages
+from swarm_copy.app.dependencies import get_session, get_user_id
+
+logger = logging.getLogger(__name__)
+
+router = APIRouter(prefix="/tools", tags=["Tool's CRUD"])
+
+
+@router.get("/{thread_id}/{message_id}")
+async def get_tool_calls(
+    session: Annotated[AsyncSession, Depends(get_session)],
+    user_id: Annotated[str, Depends(get_user_id)],
+    thread_id: str,
+    message_id: str,
+) -> list[ToolCallSchema]:
+    """Get tool calls of a specific message."""
+    await get_thread(user_id=user_id, thread_id=thread_id, session=session)
+
+    messages_result = await session.execute(
+        select(Messages)
+        .where(Messages.thread.has(user_id=user_id), Messages.thread_id == thread_id)
+        .order_by(Messages.order)
+    )
+    db_messages = messages_result.scalars().all()
+
+    # Find the message of interest.
+    try:
+        relevant_message = next(
+            (
+                i
+                for i, message in enumerate(db_messages)
+                if message.message_id == message_id
+            )
+        )
+    except StopIteration:
+        raise HTTPException(
+            status_code=404,
+            detail={
+                "detail": "Message not found.",
+            },
+        )
```

Comment:
Yes we could, my idea was that since we need all messages to do the rest of the post processing, I did not want to query the DB twice. 
Then indeed if we call the endpoint with a `human` message we get all messages for nothing.

Reply:
Fixed